File: NextcloudTalk/Notifications/NCNotification.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCNotification.h"

#import "NextcloudTalk-Swift.h"

@implementation NCNotification

+ (instancetype)notificationWithDictionary:(NSDictionary *)notificationDict
{
    if (!notificationDict || ![notificationDict isKindOfClass:[NSDictionary class]]) {
        return nil;
    }
    
    NCNotification *notification = [[NCNotification alloc] init];
    notification.notificationId = [[notificationDict objectForKey:@"notification_id"] integerValue];
    notification.app = [notificationDict objectForKey:@"app"];
    notification.objectId = [notificationDict objectForKey:@"object_id"];
    notification.objectType = [notificationDict objectForKey:@"object_type"];
    notification.subject = [notificationDict objectForKey:@"subject"];
    notification.subjectRich = [notificationDict objectForKey:@"subjectRich"];
    notification.subjectRichParameters = [notificationDict objectForKey:@"subjectRichParameters"];
    notification.message = [notificationDict objectForKey:@"message"];
    notification.messageRich = [notificationDict objectForKey:@"messageRich"];
    notification.messageRichParameters = [notificationDict objectForKey:@"messageRichParameters"];
    notification.actions = [notificationDict objectForKey:@"actions"];

    NSISO8601DateFormatter *formatter = [[NSISO8601DateFormatter alloc] init];
    notification.datetime = [formatter dateFromString:[notificationDict objectForKey:@"datetime"]];

    if (![notification.subjectRichParameters isKindOfClass:[NSDictionary class]]) {
        notification.subjectRichParameters = @{};
    }
    
    if (![notification.messageRichParameters isKindOfClass:[NSDictionary class]]) {
        notification.messageRichParameters = @{};
    }
    
    return notification;
}

- (NCNotificationType)notificationType
{
    NCNotificationType type = kNCNotificationTypeRoom;
    if ([_objectType isEqualToString:@"chat"]) {
        type = kNCNotificationTypeChat;
    } else if ([_objectType isEqualToString:@"recording"]) {
        type = kNCNotificationTypeRecording;
    } else if ([_objectType isEqualToString:@"call"]) {
        type = kNCNotificationTypeCall;
    } else if ([_objectType isEqualToString:@"remote_talk_share"]) {
        type = kNCNotificationTypeFederation;
    }
    return type;
}

- (NSString *)roomToken
{
    // Starting with NC 24 objectId additionally contains the messageId: "{roomToken}/{messageId}"
    if ([_objectId containsString:@"/"]) {
        NSArray *objectIdComponents = [_objectId componentsSeparatedByString:@"/"];
        return objectIdComponents[0];
    }

    return _objectId;
}

- (NSString *)chatMessageAuthor
{
    NSString *author = [[_subjectRichParameters objectForKey:@"user"] objectForKey:@"name"];
    NSString *guest = [[_subjectRichParameters objectForKey:@"guest"] objectForKey:@"name"];
    if (guest) {
        author = [NSString stringWithFormat:@"%@ (%@)", guest, NSLocalizedString(@"guest", nil)];
    }
    return author ? author : NSLocalizedString(@"Guest", nil);
}

- (NSString *)chatMessageTitle
{
    NSString *title = [self chatMessageAuthor];
    // Check if the room has a name
    NSArray *parameters = [self getParametersFromRichText:_subjectRich];
    for (int i = 0; i < parameters.count; i++) {
        NSTextCheckingResult *match = [parameters objectAtIndex:i];
        NSString* parameter = [_subjectRich substringWithRange:match.range];
        NSString *parameterKey = [[parameter stringByReplacingOccurrencesOfString:@"{" withString:@""]
                                  stringByReplacingOccurrencesOfString:@"}" withString:@""];
        if ([parameterKey isEqualToString:@"reaction"]) {
            return _subject;
        }
        if ([parameterKey isEqualToString:@"call"]) {
            NSString *inString = NSLocalizedString(@"in", nil);
            title = [title stringByAppendingString:[NSString stringWithFormat:@" %@ %@", inString, [[_subjectRichParameters objectForKey:@"call"] objectForKey:@"name"]]];
        }
    }
    return title;
}

- (NSArray *)getParametersFromRichText:(NSString *)text
{
    NSError *error = nil;
    NSRegularExpression *parameterRegex = [NSRegularExpression regularExpressionWithPattern:@"\\{([^}]+)\\}" options:NSRegularExpressionCaseInsensitive error:&error];
    return [parameterRegex matchesInString:text options:0 range:NSMakeRange(0, [text length])];
}

- (NSArray *)notificationActions
{
    if (!self.actions) {
        return nil;
    }

    NSMutableArray *resultActions = [[NSMutableArray alloc] init];

    for (NSDictionary *dict in self.actions) {
        NCNotificationAction *notificationAction = [[NCNotificationAction alloc] initWithDictionary:dict];

        [resultActions addObject:notificationAction];
    }

    return resultActions;
}

@end

File: NextcloudTalk/Notifications/NCNotification.swift
//
// SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

@objcMembers
class NCNotification: NSObject {

    public var notificationId: Int
    public var app: String?
    public var objectId: String
    public var objectType: String
    public var subject: String
    public var subjectRich: String
    public var subjectRichParameters: [AnyHashable: Any]
    public var message: String
    public var messageRich: String
    public var messageRichParameters: [AnyHashable: Any]
    public var actions: [[String: Any]]
    public var datetime: Date?

    init?(dictionary dict: [String: Any]?) {
        guard let dict else { return nil }

        self.notificationId = dict["notification_id"] as? Int ?? 0
        self.app = dict["app"] as? String
        self.objectId = dict["object_id"] as? String ?? ""
        self.objectType = dict["object_type"] as? String ?? ""
        self.subject = dict["subject"] as? String ?? ""
        self.subjectRich = dict["subjectRich"] as? String ?? ""
        self.message = dict["message"] as? String ?? ""
        self.messageRich = dict["messageRich"] as? String ?? ""
        self.actions = dict["actions"] as? [[String: Any]] ?? [[:]]

        self.subjectRichParameters = dict["subjectRichParameters"] as? [AnyHashable: Any] ?? [:]
        self.messageRichParameters = dict["messageRichParameters"] as? [AnyHashable: Any] ?? [:]

        if let datetime = dict["datetime"] as? String {
            let formatter = ISO8601DateFormatter()
            self.datetime = formatter.date(from: datetime)
        }
    }

    public var notificationType: NCNotificationType {
        switch objectType {
        case "chat":
            return .chat
        case "recording":
            return .recording
        case "call":
            return .call
        case "remote_talk_share":
            return .federation
        default:
            return .room
        }
    }

    public var chatMessageAuthor: String {
        if let userDict = subjectRichParameters["user"] as? [AnyHashable: Any], let userName = userDict["name"] as? String {
            return userName
        }

        if let guestDict = subjectRichParameters["guest"] as? [AnyHashable: Any], let guestName = guestDict["name"] as? String {
            return "\(guestName) (\(NSLocalizedString("guest", comment: "")))"
        }

        return NSLocalizedString("Guest", comment: "")
    }

    public var chatMessageTitle: String {
        var title = self.chatMessageAuthor

        // Check if the room has a name
        for match in self.subjectRich.ranges(of: /\{([^}]+)\}/) {
            let parameterKey = self.subjectRich[match].replacingOccurrences(of: "{", with: "").replacingOccurrences(of: "}", with: "")

            if parameterKey == "reaction" {
                return self.subject
            }

            if parameterKey == "call",
                let subjectDict = self.subjectRichParameters["call"] as? [AnyHashable: Any],
                let callName = subjectDict["name"] as? String {

                let inString = NSLocalizedString("in", comment: "")
                title = "\(title) \(inString) \(callName)"
            }

        }

        return title
    }

    public var roomToken: String {
        // Starting with NC 24 objectId additionally contains the messageId: "{roomToken}/{messageId}"
        return objectId.components(separatedBy: "/").first ?? objectId
    }

    public var threadId: Int {
        // Starting with Talk 22 objectId additionally contains the threadId: "{roomToken}/{messageId}/{threadId}"
        let components = objectId.components(separatedBy: "/")

        if components.count > 2 {
            return Int(components[2]) ?? -1
        }

        return -1
    }

    public var notificationActions: [NCNotificationAction] {
        return actions.compactMap { NCNotificationAction(dictionary: $0) }
    }
}
